producers/prometheus/prometheus.go

```diff
@@ -15,17 +15,261 @@
+			continue
+		}
+		dims := dimsToMap(message.Dimensions)
+
+		for _, d := range message.Datapoints {
+			promLog.Debugf("Processing datapoint %s", d.Name)
+			var tagKeys []string
+			var tagVals []string
+			for k, v := range dims {
+				tagKeys = append(tagKeys, sanitizeName(k))
+				tagVals = append(tagVals, v)
+			}
+			for k, v := range d.Tags {
+				tagKeys = append(tagKeys, sanitizeName(k))
+				tagVals = append(tagVals, v)
+			}
+
+			name := sanitizeName(d.Name)
+			val, err := coerceToFloat(d.Value)
+			if err != nil {
+				promLog.Warnf("Bad datapoint value %q: %s", d.Value, err)
+				continue
+			}
+			desc := prometheus.NewDesc(name, "DC/OS Metrics Datapoint", tagKeys, nil)
+			metric, err := prometheus.NewConstMetric(desc, prometheus.GaugeValue, val, tagVals...)
+			if err != nil {
+				promLog.Warnf("Could not create Prometheus metric %s", name)
+				continue
+			}
+
+			promLog.Debugf("Emitting datapoint %s", name)
+			ch <- metric
+		}
+
+	}
+}
+
+// - helpers -
+
+// writeObjectToStore writes a prefixed datapoint into the store.
+func (p *promProducer) writeObjectToStore(d producers.Datapoint, m producers.MetricsMessage, prefix string) {
+	newMessage := producers.MetricsMessage{
+		Name:       m.Name,
+		Datapoints: []producers.Datapoint{d},
+		Dimensions: m.Dimensions,
+		Timestamp:  m.Timestamp,
+	}
+	// e.g. dcos.metrics.app.[ContainerId].kafka.server.ReplicaFetcherManager.MaxLag
+	qualifiedName := joinMetricName(prefix, d.Name)
+	for _, pair := range prodHelpers.SortTags(d.Tags) {
+		k, v := pair[0], pair[1]
+		// e.g. dcos.metrics.node.[MesosId].network.out.errors.#interface:eth0
+		serializedTag := fmt.Sprintf("#%s:%s", k, v)
+		qualifiedName = joinMetricName(qualifiedName, serializedTag)
+	}
+	p.store.Set(qualifiedName, newMessage)
+}
+
+// joinMetricName concatenates its arguments using the metric name separator
+func joinMetricName(segments ...string) string {
+	return strings.Join(segments, producers.MetricNamespaceSep)
+}
+
+// janitor analyzes the objects in the store and removes stale objects. An
+// object is considered stale when the top-level timestamp of its MetricsMessage
+// has exceeded the CacheExpiry, which is calculated as a multiple of the
+// collector's polling period. This function should be run in its own goroutine.
+func (p *promProducer) janitor() {
+	ticker := time.NewTicker(p.janitorRunInterval)
+	for {
+		select {
+		case _ = <-ticker.C:
+			for k, v := range p.store.Objects() {
+				o := v.(producers.MetricsMessage)
+
+				age := time.Now().Sub(time.Unix(o.Timestamp, 0))
+				if age > p.config.CacheExpiry {
+					promLog.Debugf("Removing stale object %s; last updated %d seconds ago", k, age/time.Second)
+					p.store.Delete(k)
+				}
+			}
+		}
+	}
+}
+
+// sanitizeName returns a metric or label name which is safe for use
+// in prometheus output
+func sanitizeName(name string) string {
+	return strings.ToLower(illegalChars.ReplaceAllString(name, "_"))
```

Comment:
this is my least favorite 'feature' of prometheus

Reply:
saaaaame